kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/EnvironmentAndFacade.kt

```diff
@@ -8,30 +8,48 @@ import org.jetbrains.kotlin.cli.common.messages.CompilerMessageSourceLocation
 import org.jetbrains.kotlin.cli.common.messages.MessageCollector
 import org.jetbrains.kotlin.cli.common.messages.MessageRenderer
 import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import java.io.Closeable
+import java.io.File
 
 internal fun createEnvironmentAndFacade(
     logger: DokkaLogger,
     sourceSets: List<DokkaConfiguration.DokkaSourceSet>,
     sourceSet: DokkaConfiguration.DokkaSourceSet,
     analysisConfiguration: DokkaAnalysisConfiguration
-): EnvironmentAndFacade =
-    AnalysisEnvironment(DokkaMessageCollector(logger), sourceSet.analysisPlatform).run {
-        if (analysisPlatform == Platform.jvm) {
-            configureJdkClasspathRoots()
-        }
+): EnvironmentAndFacade {
+    val parentSourceSets = sourceSets.filter { it.sourceSetID in sourceSet.dependentSourceSets }
+    val classpath = sourceSet.classpath + parentSourceSets.flatMap { it.classpath }
+    val sources = sourceSet.sourceRoots + parentSourceSets.flatMap { it.sourceRoots }
 
-        val parentSourceSets = sourceSets.filter { it.sourceSetID in sourceSet.dependentSourceSets }
-        addClasspath(sourceSet.classpath + parentSourceSets.flatMap { it.classpath })
+    return createEnvironmentAndFacade(
+        logger = logger,
+        classpath = classpath,
+        sourceRoots = sources,
+        sourceSet = sourceSet,
+        analysisConfiguration = analysisConfiguration
+    )
+}
 
-        addSources(sourceSet.sourceRoots + parentSourceSets.flatMap { it.sourceRoots })
+internal fun createEnvironmentAndFacade(
+    logger: DokkaLogger,
+    classpath: List<File>,
+    sourceRoots: Set<File>,
+    sourceSet: DokkaConfiguration.DokkaSourceSet,
+    analysisConfiguration: DokkaAnalysisConfiguration
+) = AnalysisEnvironment(DokkaMessageCollector(logger), sourceSet.analysisPlatform).run {
+    if (analysisPlatform == Platform.jvm) {
+        configureJdkClasspathRoots()
+    }
+    addClasspath(classpath)
+    addSources(sourceRoots)
 
-        loadLanguageVersionSettings(sourceSet.languageVersion, sourceSet.apiVersion)
+    loadLanguageVersionSettings(sourceSet.languageVersion, sourceSet.apiVersion)
 
-        val environment = createCoreEnvironment()
+    val environment = createCoreEnvironment()
 
-        val (facade, _) = createResolutionFacade(environment, analysisConfiguration.ignoreCommonBuiltIns)
-        EnvironmentAndFacade(environment, facade)
-    }
+    val (facade, _) = createResolutionFacade(environment, analysisConfiguration.ignoreCommonBuiltIns)
+    EnvironmentAndFacade(environment, facade, this)
+}
```

Comment:
The very first silly thought I had was that `this` was somehow leaking through a lambda passed to the constructor

It seems like within the same `run` block there's configuration of `AnalysisEnvironment` and construction of objects for `EnvironmentAndFacade` with omitted return - I guess that confused me. Let's separate the two?

Something like that
```kotlin
internal fun createEnvironmentAndFacade(
    logger: DokkaLogger,
    classpath: List<File>,
    sourceRoots: Set<File>,
    sourceSet: DokkaConfiguration.DokkaSourceSet,
    analysisConfiguration: DokkaAnalysisConfiguration
): EnvironmentAndFacade {

    val analysisEnvironment = AnalysisEnvironment(
        messageCollector = DokkaMessageCollector(logger),
        analysisPlatform = sourceSet.analysisPlatform
    ).apply {
        if (analysisPlatform == Platform.jvm) {
            configureJdkClasspathRoots()
        }
        addClasspath(classpath)
        addSources(sourceRoots)
        loadLanguageVersionSettings(sourceSet.languageVersion, sourceSet.apiVersion)
    }

    val environment = analysisEnvironment.createCoreEnvironment()
    val (facade, _) = analysisEnvironment.createResolutionFacade(environment, analysisConfiguration.ignoreCommonBuiltIns)

    return EnvironmentAndFacade(environment, facade, analysisEnvironment)
}
```